profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp

```diff
@@ -0,0 +1,483 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#include "HeapSnapshotManager.h"
+#include "INativeThreadList.h"
+#include "OpSysTools.h"
+#include "ThreadsCpuManager.h"
+
+#include "Log.h"
+
+constexpr const WCHAR* ThreadName = WStr("DD_HeapSnapMgr");
+
+HeapSnapshotManager::HeapSnapshotManager(
+    IConfiguration* pConfiguration,
+    ICorProfilerInfo12* pCorProfilerInfo,
+    IFrameStore* pFrameStore,
+    IThreadsCpuManager* pThreadsCpuManager,
+    MetricsRegistry& metricsRegistry,
+    INativeThreadList* pNativeThreadList) :
+    ServiceBase(),
+    _inducedGCNumber(-1),
+    _session(0),
+    _gen2Size(0),
+    _lohSize(0),
+    _pohSize(0),
+    _memPressure(0),
+    _isHeapDumpInProgress(false),
+    _pCorProfilerInfo{pCorProfilerInfo},
+    _pFrameStore{pFrameStore},
+    _pThreadsCpuManager{pThreadsCpuManager},
+    _pNativeThreadList{pNativeThreadList},
+    _runtimeSessionKeywords(0),
+    _runtimeSessionVerbosity(0),
+    _startTimestamp(0ns),
+    _lastTimestamp(0ns),
+    _lastOldHeapSize(0),
+    _lastMemPressure(0),
+    _shouldStartHeapDump(false),
+    _shouldCleanupHeapDumpSession(false),
+    _loopThreadOsId(0),
+    _objectCount(0),
+    _totalSize(0),
+    _duration(0)
+{
+    _heapDumpInterval = pConfiguration->GetHeapSnapshotInterval();
+    _memPressureThreshold = pConfiguration->GetHeapSnapshotMemoryPressureThreshold();
+    _snapshotCheckInterval = pConfiguration->GetHeapSnapshotCheckInterval();
+
+    _heapSnapshotDurationMetric = metricsRegistry.GetOrRegister<ProxyMetric>("dotnet_heapsnapshot_duration", [this]() {
+        return static_cast<double>(_duration);
+    });
+
+    _heapSnapshotObjectCountMetric = metricsRegistry.GetOrRegister<ProxyMetric>("dotnet_heapsnapshot_object_count", [this]() {
+        return static_cast<double>(_objectCount);
+    });
+
+    _heapSnapshotTotalSizeMetric = metricsRegistry.GetOrRegister<ProxyMetric>("dotnet_heapsnapshot_total_size", [this]() {
+        return static_cast<double>(_totalSize);
+    });
+
+
+    _pCorProfilerInfo->AddRef();
+}
+
+HeapSnapshotManager::~HeapSnapshotManager()
+{
+    StopImpl();
```

Comment:
shouldn't we call `Stop` instead of `StopImpl`.
`Stop` will modify the state of the service but not `StopImpl`. 
In this case it's not big, because it lives as long as the process lives, but just to be clean